secret_store/src/key_server_cluster/admin_sessions/servers_set_change_session.rs

```diff
@@ -1045,148 +1045,205 @@ fn check_nodes_set(all_nodes_set: &BTreeSet<NodeId>, new_nodes_set: &BTreeSet<No
 pub mod tests {
 	use std::sync::Arc;
 	use std::collections::{VecDeque, BTreeMap, BTreeSet};
+	use ethereum_types::H256;
 	use ethkey::{Random, Generator, Public, Signature, KeyPair, sign};
-	use key_server_cluster::{NodeId, SessionId, Error, KeyStorage, DummyKeyStorage};
-	use key_server_cluster::cluster::Cluster;
+	use key_server_cluster::{NodeId, SessionId, Error, KeyStorage, NodeKeyPair, PlainNodeKeyPair};
 	use key_server_cluster::cluster_sessions::ClusterSession;
-	use key_server_cluster::cluster::tests::DummyCluster;
-	use key_server_cluster::generation_session::tests::{MessageLoop as GenerationMessageLoop, Node as GenerationNode, generate_nodes_ids};
+	use key_server_cluster::cluster::tests::MessagesLoop;
+	use key_server_cluster::generation_session::tests::{MessageLoop as GenerationMessageLoop};
+	use key_server_cluster::math;
 	use key_server_cluster::message::Message;
 	use key_server_cluster::admin_sessions::ShareChangeSessionMeta;
-	use key_server_cluster::admin_sessions::share_add_session::tests::check_secret_is_preserved;
 	use key_server_cluster::jobs::servers_set_change_access_job::ordered_nodes_hash;
 	use super::{SessionImpl, SessionParams};
 
-	struct Node {
-		pub cluster: Arc<DummyCluster>,
-		pub key_storage: Arc<DummyKeyStorage>,
-		pub session: SessionImpl,
+	pub trait AdminSessionAdapter<S> {
+		const SIGN_NEW_NODES: bool;
+
+		fn create(
+			meta: ShareChangeSessionMeta,
+			admin_public: Public,
+			all_nodes_set: BTreeSet<NodeId>,
+			ml: &MessagesLoop,
+			idx: usize
+		) -> S;
 	}
 
-	struct MessageLoop {
+	pub struct MessageLoop<S> {
+		pub ml: MessagesLoop,
```

Comment:
hrm, am I reading this correctly: is every MessageLoop instance containing another MessageLoop?